Add Read Update Delete tests for ThreadPools and an end-to-end integration test

lloydmeta opened the pull request:
- For admin ThreadPools, During the 2.4.x migration, the Create integration test was already added.
- This PR also adds an end-to-end integration test suite (listing Parts, asking for parts)

lloydmeta committed:
Add Read Update Delete tests for ThreadPools

During the 2.4.x migration, the Create test was already added

lloydmeta committed:
Add a test to ensure ThreadPools that are assigned to a part do not get deleted

lloydmeta committed:
Add End to end integration test

The main point of this commit is to add an end-to-end integration test that checks that the OctoClient can retrieve a list of Parts and
can request information by making PartRequests to defined and undefined Parts.

Other changes:

- Add OctoMetricsImpl, an implementation of Kenshoo.Metrics that attempts to clean up the registry before proceeding with instantiation
  - This prevents tests from occasionally failing due to asynchronous cleanup
- Use aforementioned newly created MetricsImpl
- Use contains instead of == Some(thing) in PartsController to check for cache control header
- Fix harness-related Spec traits
  - Introduce PlayAppComponents trait to hold Play application pieces but does not re-define run method
  - RequiresDB no longer extends PlayAppSupport directly (because the latter defines its own run method)
  - PlayServerSupport no longer extends PlayAppSupport in order to prevent Play.start(app) from being called twice and causing shutdown of components before they're used (see Scaladoc for more info)

xevix commented on project/Octoparts.scala:
```diff
@@ -20,7 +20,7 @@ object Octoparts extends Build {
   lazy val app = Project(id = "octoparts", base = file("."), settings = playAppSettings)
     .enablePlugins(PlayScala)
-    .dependsOn(models, authHandlerApi, playJsonFormats)
+    .dependsOn(models, authHandlerApi, playJsonFormats, scalaWsClient % "test->compile")
```
Could you explain this particular change?

lloydmeta replied:
Yeah, this is a cryptic addition that adds a test-mode dependency in the current project definition (root app in this case); in particular the ScalaWsClient project is to be in a compiled state before running tests in root.

The reason I did this was to allow us to integration test the root project (Octoparts app) using the ScalaWsClient.

xevix replied:
Ah I see. Maybe a one-liner comment to explain this would be good then.

lloydmeta replied:
Good idea. Done :+1:

xevix commented on test/integration/AdminSpec.scala:
```diff
@@ -18,15 +18,100 @@ class AdminSpec
+    describe("the show Thread Pool page") {
+
+      it("should redirect me to the thread poollist page if no such thread pool can be found") {
+        goTo(ThreadPoolShowPage(35415151413L))
```
Is this just an arbitrarily-big Long?

lloydmeta replied:
Yes.

lloydmeta replied:
After discussion, will use `Long.MaxValue`

lloydmeta replied:
Done.

lloydmeta committed:
Replace arbitrary Longs with Long.MaxValue

lloydmeta committed:
Add a comment explaining test->compile

xevix commented on test/integration/AdminSpec.scala:
```diff
@@ -36,7 +121,7 @@ class AdminSpec
 
       it("should work and redirect me to the part show page") {
-        val PartAddPage.PartConfig(partId, url, connectionPoolSize, commandKey, commandKeyGroup, proxy) = PartAddPage.createPart(ThreadPoolAddPage.createThreadPool)
+        val PartAddPage.PartConfig(partId, url, connectionPoolSize, commandKey, commandKeyGroup, Some(proxy)) = PartAddPage.createPart(ThreadPoolAddPage.createThreadPool._2)
```
Perhaps split the tuple into its own line for clarity, as `._2` is not immediately obvious.

``` scala
val (_, threadPoolConfig) = ThreadPoolAddPage.createThreadPool
val PartAddPage.PartConfig(partId, _, _, _, _, _) = PartAddPage.createPart(threadPoolConfig)
```

Likewise for L143.

lloydmeta replied:
Good idea. Done.

lloydmeta committed:
Assign thread pool to an intermediary val for clarity

xevix commented on app/com/m3/octoparts/util/OctoMetricsImpl.scala:
```diff
@@ -0,0 +1,27 @@
+ */
+class OctoMetricsImpl(lifecycle: ApplicationLifecycle, configuration: Configuration) extends MetricsImpl(lifecycle, configuration) {
+
+  override def onStart(): ObjectMapper = {
```
Is it possible to add a test for this, or is it too edge-casey? I suppose you'd just be testing to see if the metric was removed from the registry.

Although the comment above mentions that this is to address issues when running tests, this implementation with the override is being used in non-test code (`UtilsModule`) in place of the default `MetricsImpl`. In this case, I think that either this needs testing, or that the custom implementation should only be used in testing.

lloydmeta replied:
I think you're right in saying that this custom implementation should be tested in order to make sure it does what it is supposed to do: remove metrics with the same names from the registry when `onStart` is run and _then_ go about adding metrics into the registry. I think it should be enough to add a test that does the following:
1. Sets up the global registry singleton to have the said metrics
2. Instantiates OctoMetricsImpl
3. Calls onStart
4. Ensures the metrics are in the registry
   - Implicitly, we know the custom implementation works because no error has been thrown during 3.

What do you think?

xevix replied:
Calling `onStart` attempts to remove the metric from the internal map. I'm guessing the error thrown would be from the concurrent map [in use here in Codahale Metrics](https://github.com/dropwizard/metrics/blob/d8d8e2796d1f7dfb5268e2e8fbf35039570c53d7/metrics-core/src/main/java/io/dropwizard/metrics/SharedMetricRegistries.java#L26) itself [called from Kenshoo Metrics](https://github.com/kenshoo/metrics-play/blob/master/src/main/scala/com/kenshoo/play/metrics/Metrics.scala#L80). What would it throw specifically, that it could not remove a key not in the map?

lloydmeta replied:
`onStart` with this custom implementation (`OctoMetricsImpl`) will try to remove the metrics from the internal metrics registry map by invoking the parent `onStop` cleanup method before proceeding with normal `onStart` duties; therefore, should be no error thrown; still, a good idea to catch NonFatal errors.

The default implementation of `onStart` in `MetricsImpl` will not try to clean up first; it will therefore throw an error if there are already metrics registered in the global singleton with the same names as the ones that it tries to register (`IllegalArgumentException`)

xevix replied:
Ah ok, so the catch here was more for paranoia (which I'm totally for when it comes to internal config issues).

Yeah, this plan sounds sound (ugh).

lloydmeta replied:
Done in af1f508

lloydmeta committed:
Add a testsuite for OctoMetricsImpl

xevix commented on test/com/m3/octoparts/util/OctoMetricsImplSpec.scala:
```diff
@@ -0,0 +1,44 @@
+      subject.onStart()
+    }
+
+    it("should work with consecutive calls") {
```
I'm bordering on pedantry, but does "should work" mean "should not throw" in this context?

lloydmeta replied:
Yes. Renamed for clarity.

xevix commented on test/com/m3/octoparts/util/OctoMetricsImplSpec.scala:
```diff
@@ -0,0 +1,44 @@
+      subject.onStart()
+    }
+
+    it("should work if metrics were already registered") {
```
More pedantry and my own confirmation, this is checking if the _same_ metrics were already registered on the same default registry, yes?

A slight suggestion would be to create the subject with `jvmMetricsEnabled` true, and just call `setupJvmMetrics` before the `onStart` call. Unfortunately the specific attributes that get enabled aren't factored into a list in Kenshoo so the `find`s will have to remain =/

lloydmeta replied:
The original intent was to make it obvious within the test that _these_ are the metrics we want to ensure are registered before calling `.onStart`, and to make that expectation explicit.

However, thinking about it again, it's just as easy to make it explicit by refactoring out the check into a separate method, call that method after calling `setupJvmMetrics`, and then call it again after calling `.onStart`. Done as part of 7a1e4fc

xevix replied:
Great! I hope this didn't sound too pedantic, thanks for the refactor.

lloydmeta committed:
- Use setupJvmMetrics
- Introduce helper function for ensuring metric names that we want are defined
- Rename test case for clarity

xevix commented:
LGTM. Does this merit a version bump and a release?

lloydmeta commented:
Nah, I don't think so; [this line that modifies the header check](https://github.com/m3dev/octoparts/pull/192/files#diff-92f67571d02b81ad83359da68946b9eeR67) is the only impactful change in running-code and I doubt we'll see noticeable performance improvements from that :laughing:
